Update Selenium Manager (beta) page (#1454)
SeleniumHQ/seleniumhq.github.io: website_and_docs/content/documentation/selenium_manager.en.md

linkTitle: "Selenium Manager"
weight: 11
description: >

>     Selenium Manager is a binary tool implemented in Rust that provides automated driver and browser management for Selenium. Selenium bindings use this tool by default, so you do not need to download it or add anything to your code or do anything else to use it.

Reviewer: ```suggestion
    Selenium Manager is a command-line tool implemented in Rust that provides automated driver and browser management for Selenium. Selenium bindings use this tool by default, so you do not need to download it or add anything to your code or do anything else to use it.
```

---

## Motivation

> **TL;DR**: **Selenium Manager** is the official *driver manager* of the Selenium project, and it is shipped out of the box with every Selenium release as of version 4.6.

Reviewer: ```suggestion
**TL;DR**: **Selenium Manager** is the official *driver manager* of the Selenium project, and it is shipped out of the box with every Selenium release as of version 4.6. Some features are unreliable before Selenium 4.11, though, so please use the latest version of Selenium to get the most use out of this feature.
```

Author: I prefer to avoid this sentence being added in the TLDR. IMO it's evident that things become more stable with more recent releases.

Reviewer: Can we remove the reference to 4.6, then? We actually don't want people to use 4.6, so unless we're doing a version history recap I don't think mentioning it helps.

Author: Done.

> Selenium uses the native support implemented by each browser to carry out the automation process. For this reason, Selenium users need to place a component called _driver_ (e.g., chromedriver, geckodriver, msedgedriver, etc.) between the script using the Selenium API and the browser. For many years, managing these drivers was a manual process for Selenium users. This way, they had to download the required driver for a browser (e.g., chromedriver for Chrome, geckodriver for Firefox, etc.) and place it in the `PATH` or export the driver path as a system property (e.g., Java, JavaScript, etc.). But this process was cumbersome and led to maintainability issues.

Reviewer: nit: use either "e.g.," or "etc." — the statement stops making sense when using both.

Author: Will change.

Author: Done.

This problem is the primary reason for the existence of the so-called *driver managers* (such as [WebDriverManager](https://bonigarcia.dev/webdrivermanager/) for Java, 
[webdriver-manager](https://pypi.org/project/webdriver-manager/) for Python, [webdriver-manager](https://www.npmjs.com/package/webdriver-manager) for JavaScript, [WebDriverManager.Net](https://github.com/rosolko/WebDriverManager.Net) for C#, and [webdrivers](https://github.com/titusfortner/webdrivers) for Ruby. All these projects were an inspiration and a clear sign that the community needed this feature to be built in Selenium. Thus, the Selenium project has created *Selenium Manager*, the official driver manager for Selenium, shipped out of the box with each Selenium release as of version 4.6.

## Usage
**TL;DR**: Selenium Manager is used by the Selenium bindings when the drivers (chromedriver, geckodriver, etc.) are unavailable.

> Driver management through Selenium Manager is *opt-in* for the Selenium bindings. Thus, users can continue managing their drivers manually (putting the driver in the `PATH` or using system properties) or rely on a third-party *driver manager* to do it automatically. Selenium Manager only operates as a fallback: if no driver is provided, Selenium Manager will come to the rescue.

Reviewer: ```suggestion
Driver management through Selenium Manager is currently *opt-in* for the Selenium bindings. Thus, users can continue managing their drivers manually (putting the driver in the `PATH` or using system properties) or rely on a third-party *driver manager* to do it automatically. Selenium Manager only operates as a fallback: if no driver is provided, Selenium Manager will come to the rescue.
```

> Selenium Manager is a CLI (command line interface) tool implemented in Rust to allow cross-platform execution and compiled for Windows, Linux, and macOS. The Selenium Manager binaries are shipped with each Selenium release. This way, each Selenium binding language invokes Selenium Manager to carry out the automated driver and browser management explained in the following sections.

Reviewer: ```suggestion
Selenium Manager is a CLI (command line interface) tool implemented in Rust to allow cross-platform execution and compiled for Windows, Linux, and macOS. The Selenium Manager binaries are shipped with each Selenium release. This way, each Selenium binding language invokes the correct Selenium Manager executable to carry out the automated driver and browser management explained in the following sections.
```

2. Driver version discovery. With the discovered browser version, the proper driver version is resolved. For this step, the online metadata/endpoints maintained by the browser vendors (e.g., [chromedriver](https://chromedriver.chromium.org/downloads), [geckodriver](https://github.com/mozilla/geckodriver/releases), or [msedgedriver](https://developer.microsoft.com/en-us/microsoft-edge/tools/webdriver/)) are used.
3. Driver download. The driver URL is obtained with the resolved driver version; with that URL, the driver artifact is downloaded, uncompressed, and stored locally.
4. Driver cache. Uncompressed driver binaries are stored in a local cache folder (`~/.cache/selenium`). The next time the same driver is required, it will be used from there if the driver is already in the cache.

## Automated browser management
**TL;DR**: Selenium Manager **automatically discovers, downloads, and caches the browsers** driven with Selenium (Chrome, Firefox, and Edge) when these browsers are not installed in the local system.

> As of Selenium 4.11.0, Selenium Manager also implements *automated browser management*. With this feature, Selenium Manager allows us to discover, download, and cache the different browser releases, making them seamlessly available for Selenium. Internally, Selenium Manager uses an equivalent management procedure explained in the section before, but this time, for browser releases.

Reviewer: ```suggestion
As of Selenium 4.11.0, Selenium Manager also implements *automated browser management* for Chrome releases 113+ with [Chrome for Testing (CfT)](https://googlechromelabs.github.io/chrome-for-testing/). With this feature, Selenium Manager allows us to discover, download, and cache the browser version requested in the Browser Options class, making them seamlessly available for Selenium. Internally, Selenium Manager uses an equivalent management procedure explained in the section before, but this time, for browser releases.
```

Author: This info is explained later, so I would not change this paragraph.

Reviewer: I think this paragraph overstates what has been implemented, and the clarification in the next paragraph ends up sounding more contradictory rather than explanatory.

I think we may be getting caught up in versioning too much? Maybe we write the whole document specifying that it applies to Selenium 4.12 and avoiding discussing when which feature was implemented (unless we want to add a separate section). Then we just note which features are still in progress.

Author: I don't know, perhaps. But I would leave it like that for now (since we are in 4.11.0 and heading to 4.12.0) and review this in a later editing round.

> The browser automatically managed by Selenium Manager are:

Reviewer: I think the CfT info belongs in previous paragraph and for now this section should just mention that Firefox and Edge are under development.

When these labels are specified, Selenium Manager first checks if a given browser is already installed (`beta`, `dev`, etc.), and when it is not detected, the browser is automatically managed.

## Configuration

> **TL;DR**: Selenium Manager should work silently and transparently for most users. Nevertheless, there are scenarios (e.g., when a proxy or firewall forbids making network requests) where custom configuration is required.

Reviewer: The bindings do support Proxies out of the box if users specify the proxy information directly in the Options class (as opposed to using args or environment variables)

Author: Changed that TLDR.

In addition to the configuration keys specified in the table before, there are some special cases, namely:

- Browser version. In addition to `browser-version`, we can use the specific configuration keys to specify custom versions per supported browser. This way, the keys `chrome-version`, `firefox-version`, `edge-version`, etc., are supported. The same applies to environment variables (i.e., `SE_CHROME_VERSION`, `SE_FIREFOX_VERSION`, `SE_EDGE_VERSION`, etc.).

> - Driver version. Following the same pattern, we can use `chromedriver-version`, `geckodriver-version`,  `msedgedriver-version`,  etc. (in the configuration file), and `SE_CHROMEDRIVER_VERSION`, `SE_GECKODRIVER_VERSION`, `SE_MSEDGEDRIVER_VERSION`,  etc. (as environment variables).

Reviewer: I kind of think we should merge browser version and driver version before we get out of beta, but that's a discussion for another time. 😄

Author: Browser versions and driver versions are different things. Maybe similar in the case of Chrome and Edge since they follow the same versioning schema for browser and driver, but not Firefox (e.g., Firefox 116 needs geckodriver 0.32.0).

> - Browser path. Following the same pattern, we can use `chrome-path`, `firefox-path`,  `edge-path`,  etc. (in the configuration file), and `SE_CHROME_PATH`, `SE_FIREFOX_PATH`, `SE_EDGE_PATH`,  etc. (as environment variables).

Reviewer: I think we want to point out that this is also available inside the bindings from the Options class with the appropriate binary method.

Author: I will add a reference to that. I think is this link, right?

https://www.selenium.dev/documentation/webdriver/browsers/edge/#start-browser-in-a-specified-location

That part is currently incomplete, unfortunately.

Reviewer: yeah, there's a different link for Chrome, Edge and Firefox (doesn't apply to IE or Safari)

Reviewer: and yeah, would be good to get those examples documented.

Author: I see. I will use Chrome's then.

Author: I added Chrome, Firefox, and Edge.

Selenium Manager follows the same versioning schema as Selenium. Nevertheless, we use the major version 0 for Selenium Manager releases because it is still in beta. For example, the Selenium Manager binaries shipped with Selenium 4.12.0 corresponds to version 0.4.12.

## Getting Selenium Manager
For most users, direct interaction with Selenium Manager is not required since the Selenium bindings use it internally. Nevertheless, if you want to *play* with Selenium Manager or use it for your use case involving driver or browser management, you can get the Selenium Manager binaries in different ways:

> - From the cache. As of Selenium 4.12.0, the Selenium Manager binaries are extracted from each binding distribution and copied to the cache folder. For instance, the Selenium Manager binary shipped with Selenium 4.12.0 will be stored in the folder `~/.cache/selenium/manager/0.4.12`).

Reviewer: This isn't going to be supported in 4.12. Too much more to add. Let's say that it's under development.

Author: That's a pity. Will change.

Reviewer: I wanted to release 4.12 2 days ago and .NET just kicked my butt. :)

Author: Changed.

```

> ## Selenium Grid

Reviewer: So, grid is weird right now because you can't use the dynamic aspect of setting the browser version in options. Right now, you need to create a stereotype with the supported versions so that the options will match the node. I think we should change this, but that's still under discussion. 😄 Regardless I think it's a caveat we should spell out.

Author: I don't know more about what's going on with Selenium Grid than I explained in this doc. Please extend/change it if required.

Reviewer: What I know is explained in this issue - https://github.com/SeleniumHQ/selenium/issues/12530
My lack of grid experience means I don't actually know how to make it work properly right now, either, is part of the problem. ☹️
